components/style/matching.rs

```diff
@@ -769,22 +769,27 @@ pub trait MatchMethods : TElement {
                 //
                 //  2. This is an incremental restyle, but the old display value
                 //     is none, so there's no effective way for Gecko to get the
-                //     style source. In this case, we could return either
+                //     style source (which is the style context).
+                //
+                //     In this case, we could return either
                 //     RestyleDamage::empty(), in the case both displays are
                 //     none, or rebuild_and_reflow, otherwise. The first case
                 //     should be already handled when calling this function, so
                 //     we can assert that the new display value is not none.
                 //
-                //     Also, this can be a text node (in which case we don't
-                //     care of watching the new display value).
+                //     Also, note that some styled pseudo-elements may not have
+                //     a frame even when the display value is other than none,
+                //     so we need to actually compare both values here.
                 //
-                // Unfortunately we can't strongly assert part of this, since
-                // we style some nodes that in Gecko never generate a frame,
-                // like children of replaced content. Arguably, we shouldn't be
-                // styling those here, but until we implement that we'll have to
-                // stick without the assertions.
-                debug_assert!(pseudo.is_none() ||
-                              new_style.get_box().clone_display() != display::T::none);
+                if let Some(old_style) = old_style {
+                    debug_assert!(pseudo.is_some() ||
+                                  old_style.get_box().clone_display() == display::T::none,
+                                  "Should've reached the frame!");
```

Comment:
I guess generating a reconstruct frame hint for these is ok until we do that. Should I remove the assertion in the Gecko case?